### DISCUSSION TITLE
how does useSelector cache without re-running the selector ?

### DISCUSSION THREAD

**Discussion by ethan-cao** [reduxjs/react-redux#1892] [Q&A] [answered]
Hi 👋

From the [doc](https://react-redux.js.org/api/hooks#equality-comparisons-and-updates), `A cached result may be returned by the hook without re-running the selector if it's the same function reference as on a previous render of the component`.

I have tried supplying `useSelecor` with the same selecor instance during re-render, however the selector still got called. 

Here is the [codesandbox](https://codesandbox.io/s/simple-react-redux-example-forked-x208ns?file=/src/App.js:0-530) 
Whenever I click the `Add one` button, selector gets called. However, I was expecting it not got called based on the doc.

Did I misunderstand something?

**Comment by markerikson** [comment, marked as answer]
It'll also re-run if the state changed as well:

https://github.com/reduxjs/react-redux/blob/9306158197afcb03ad4e2b56e53fa5115320f354/src/hooks/useSelector.js#L31-L48

**Reply by ethan-cao** [reply]
Thanks @markerikson!

I tried dispatching an random aciton which does not change the state, then the selector is indeed skiped.

**Reply by ethan-cao** [reply]
In the doc, we say `useSelector() will also subscribe to the Redux store, and run your selector whenever an action is dispatched.`

Does `whenever an action is dispatched` actually means `whenever the state changes`?

**Reply by markerikson** [reply]
Roughly, yeah.

The Redux store itself always runs all subscribers after _every_ dispatch, regardless of whether or not any state was updated. 

But, `useSelector` tries to optimize behavior - if the `rootState` hasn't changed, then it skips running the selector.

Most of the time we can shorten that and assume that "if an action was dispatched, some state got updated". But, it _is_ very possible to dispatch actions that result in no state updates.
